app-backend/tile/src/main/scala/LayerCache.scala

```diff
@@ -37,96 +40,92 @@ import scala.util._
   * things that require time to generate, usually a network fetch, use AsyncLoadingCache
   */
-object LayerCache extends Config {
+object LayerCache extends Config with LazyLogging {
   implicit val database = Database.DEFAULT
 
   val memcachedClient = KryoMemcachedClient.DEFAULT
 
   /** The caffeine cache to use for attribute stores */
-  val attributeStoreCache: ScaffeineCache[String, Future[S3AttributeStore]] =
+  private val attributeStoreCache: ScaffeineCache[String, AttributeStore] =
     Scaffeine()
       .recordStats()
       .expireAfterAccess(5.minutes)
       .maximumSize(500)
-      .build[String, Future[S3AttributeStore]]()
+      .build[String, AttributeStore]()
 
-  /**
-    * The primary means of interaction with this class: pass as key and
-    *  the costly option caching prevents duplication of.
-    */
-  def attributeStore(bucket: String, prefix: Option[String]): Future[S3AttributeStore] = {
-    val cacheKey = HeapBackedMemcachedClient.sanitizeKey(s"store-$bucket-$prefix")
-    attributeStoreCache.get(cacheKey, { cacheKey =>
-      prefix match {
-        case Some(prefixStr) => Future.successful(S3AttributeStore(bucket, prefixStr))
-        case None => Future.failed(new LayerIOError("Scene has no ingest location"))
-      }
-    })
-  }
+  private val histogramCache = HeapBackedMemcachedClient(memcachedClient)
+  private val tileCache = HeapBackedMemcachedClient(memcachedClient)
+  private val layerLocationCache = HeapBackedMemcachedClient(memcachedClient)
 
-  def attributeStore(prefix: Option[String]): Future[S3AttributeStore] =
-    attributeStore(defaultBucket, prefix)
+  def layerUri(layerId: UUID): Future[Option[String]] = {
+    layerLocationCache.caching(s"layer-uri-${layerId.toString}") { _ =>
+      Scenes.getScene(layerId).map(_.flatMap(_.ingestLocation))
+    }
+  }
 
-  val tileCache = HeapBackedMemcachedClient[Option[MultibandTile]](memcachedClient)
-  def maybeRenderExtent(id: UUID, zoom: Int, extent: Extent): Future[Option[MultibandTile]] =
-    tileCache.caching(s"rendered-extent-$id-$zoom-$extent") { implicit ec =>
-      for {
-        prefix <- prefixFromLayerId(id)
-        store <- attributeStore(defaultBucket, prefix)
-      } yield {
-        val reader = new S3ValueReader(store).reader[SpatialKey, MultibandTile](LayerId(id.toString, zoom))
-        val query = S3CollectionLayerReader(store)
-          .query[SpatialKey, MultibandTile, TileLayerMetadata[SpatialKey]](LayerId(id.toString, zoom))
-        val tile = query.where(Intersects(extent))
-          .result
-          .stitch
-          .crop(extent)
-          .tile
-        tile match {
-          case tile: MultibandTile => Some(tile)
-          case _ => None
-        }
+  def attributeStoreForLayer(layerId: UUID): Future[Option[AttributeStore]] = {
+    // we are not caching the attribute store based on layerId but based on the catalog location
+    OptionT(layerUri(layerId)).mapFilter { catalogUri =>
```

Comment:
Still new at thinking about `OptionT`, and [this](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/OptionT.scala#L34-L35), well, you know

![](http://68.media.tumblr.com/5bb9e3b21fafc55b1c615f87d8605aab/tumblr_ognd3lJaZa1tu4c7to7_400.gif)

Is herding cats a good place to look for explanations? I found, e.g., [this post](http://eed3si9n.com/herding-cats/FunctorFilter.html)